src/prolly/map.ts

```diff
@@ -1,104 +1,131 @@
+    const entries = this._entries;
+    let size = entries.length;
+    if (size === 0) {
+      return {found: false, index: 0};
+    }
+    let base = 0;
+
+    while (size > 1) {
+      const half = Math.floor(size / 2);
+      const mid = base + half;
+      // mid is always in [0, size), that means mid is >= 0 and < size.
+      // mid >= 0: by definition
+      // mid < size: mid = size / 2 + size / 4 + size / 8 ...
+      const entry = entries[mid];
+      // No way that key can be None.
+      const cmp = stringCompare(entry[0], key);
+      base = cmp > 0 ? base : mid;
+      size -= half;
+    }
+    // base is always in [0, size) because base <= mid.
+    const entry = entries[base];
+    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+    const cmp = stringCompare(entry[0], key);
+    if (cmp === 0) {
+      return {found: true, index: base};
     }
-    return this._base.binarySearch(key).found;
+    return {found: false, index: base + cmp};
   }
 
-  get(key: string): JSONValue | undefined {
-    const ks = key;
-    const p = this._pending.get(ks);
-    switch (p) {
-      case deleteSentinel:
-        return undefined;
-      case undefined:
-        return this._baseGet(key);
-      default:
-        return p;
-    }
+  has(key: string): boolean {
+    const {found} = this._binarySearch(key);
+    return found;
   }
 
-  private _baseGet(key: string): JSONValue | undefined {
-    if (this._base === undefined) {
-      return undefined;
-    }
-    const {found, index} = this._base.binarySearch(key);
+  get(key: string): ImmutableJSONValue | undefined {
+    const {found, index} = this._binarySearch(key);
     if (!found) {
       return undefined;
     }
-    return this._base.entries[index][1];
+    const [, value] = this._entries[index];
+    return value;
   }
 
   put(key: string, val: JSONValue): void {
-    this._pending.set(key, val);
+    const frozen = deepFreeze(val);
+    const {found, index} = this._binarySearch(key);
+    if (found) {
+      this._entries[index] = [key, frozen];
+    } else {
+      this._entries.splice(index, 0, [key, frozen]);
+    }
+    this._changedKeys.add(key);
   }
 
   del(key: string): void {
-    this._pending.set(key, deleteSentinel);
+    const {found, index} = this._binarySearch(key);
+    if (found) {
+      this._entries.splice(index, 1);
+    }
+    this._changedKeys.add(key);
   }
 
   entries(): IterableIterator<Entry> {
-    return new Iter(this._base, this._pending);
+    return this._entries[Symbol.iterator]();
   }
 
   [Symbol.iterator](): IterableIterator<Entry> {
     return this.entries();
   }
 
   async flush(write: dag.Write): Promise<string> {
-    const newBase = new Leaf([...this]);
-    await write.putChunk(newBase.chunk);
-    this._base = newBase;
-    this._pending.clear();
-    return this._base.chunk.hash;
+    // TODO: how come ts doesn't complain about passing the mutable array?
```

Comment:
I played around with this yesterday and it makes sense but it is not very helpful for us.

Reply:
I filed https://github.com/rocicorp/replicache/issues/483.